### DISCUSSION TITLE
CMSIS 6.1.2 refresh hangs forever in AnalogDevices

### DISCUSSION THREAD

**Discussion by jkraskesis** [eclipse-embed-cdt/eclipse-plugins#518] [Managed Build Plug-ins Q&A] [answered]
Hi,

When I refresh the CMSIS packs using Embedded C/C++ CMSIS Packs (Experimental) 6.1.2 the procedure hangs forever while accessing AnalogDevices ADuCM320_DFP (from https://www.keil.com/pack/index.pidx:  <pdsc url="http://www.analog.com/media/en/engineering-tools/design-tools" vendor="AnalogDevices" name="ADuCM320_DFP" version="1.1.0" />). Accessing the url with a browser give "no such page".

Is there any other way to get the Keil.STM32F1xx_DFP.2.4.0.pack available in Eclipse?

TIA

Joost

**Comment by ilg-ul** [comment]
At this moment the url seems functional. It might have been a temporary problem with the server.

**Comment by jkraskesis** [comment]
No, it is not from here: the "Progress Information"-box hangs on "AnalogDevices.ADuCM320_DFP.pdsc". And it is not possible to cancel the operation. I have to "kill" Eclipse completely.

BTW: with accessing the URL with a browser I meant accessing "http://www.analog.com/media/en/engineering-tools/design-tools".

**Comment by ilg-ul** [comment]
Please provide a window capture with the error. The message box should have a button to skip faulty urls.

Also please copy/paste the errors from the console.

**Reply by jkraskesis** [reply]
There is no error. The "Progress Information"-box just hangs. It does not have a button to skip faulty URLS. During the refresh procedure there were some faulty urls and indeed, they had an "Ignore" and "Ignore all" button.

![Screenshot](https://user-images.githubusercontent.com/28053200/173192172-d0ee82fc-6770-4fb4-a942-3d8bb8fe2a34.png)

The console output from Eclipse on Debian Bookworm with openjdk 17.0.3 2022-04-19:

jkr@panoramix:~/disks/datadisk2/EclipseTest/eclipse-embedcpp-2022-03-R-linux-gtk-x86_64/eclipse$ ./eclipse 
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.



(Eclipse:4080778): Gtk-CRITICAL **: 16:29:01.260: gtk_box_gadget_distribute: assertion 'size >= 0' failed in GtkScrollbar

(Eclipse:4080778): Gtk-CRITICAL **: 16:29:01.296: gtk_box_gadget_distribute: assertion 'size >= 0' failed in GtkScrollbar

(Eclipse:4080778): Gtk-CRITICAL **: 16:29:01.303: gtk_box_gadget_distribute: assertion 'size >= 0' failed in GtkScrollbar

**Reply by ilg-ul** [reply]
> ... The console output from Eclipse on Debian Bookworm with openjdk 17.0.3 2022-04-19

Sorry for the confusion, I mean the packs console, it is a separate Eclipse tab where the cmsis packs plug-in logs its activity.

**Reply by jkraskesis** [reply]
Ah, see the screenshot. As Eclipse does not react I cannot provide the complete console output
![Screenshot](https://user-images.githubusercontent.com/28053200/173195265-94380922-3b95-48a9-8c35-12d88cf2fc6a.png)
.

**Reply by ilg-ul** [reply]
~Unfortunately I cannot reproduce the issue, on my Mac the **Update the CMSIS packs configuration** was fine, it took a few minutes but did not hang.~

~It is a simple guess, but it might be something related to security settings, probably one of the urls require some configuration that the Java library does not support properly.~

~I am not familiar with the new JDK features/issues. :-(~

**Reply by jkraskesis** [reply]
Can you give me the complete url that the plugin actually uses to get the content from the "http://www.analog.com/media/en/engineering-tools/design-tools" website? Maybe I can debug it with that, using other tools?

**Reply by ilg-ul** [reply]
The plug-in gets all urls from the index you mentioned at the beginning.

**Reply by ilg-ul** [reply]
I'm not sure, but the url you are looking for seems to be:

- <http://www.analog.com/media/en/engineering-tools/design-tools/ADuCM320_DFP.pdsc>

On my machine, trying to download it with the browser brings a _Page Not Found_ html file.

Probably on your machine something else happens, which confuses Java.

These kind of problems are quite common with CMSIS Packs, vendors submit a URL to Keil, then something weird happens and the Keil index gets out of sync. This is one of the reasons I gave up on using CMSIS Packs and created xPacks.

**Comment by ilg-ul** [comment]
After some fiddling, I managed to reproduce the issue. The reason it did not show before was the cache; after removing it, access to that url hanged, as you described.

I'll further investigate and try to find a fix.

**Reply by ilg-ul** [reply]
The AnalogDevices server, instead of returning the file or a HTTP error, returns a long HTML with lots of scripts.

Apparently the Java XML parser is not happy with the HTML and gets confused. :-(

**Comment by ilg-ul** [comment, marked as answer]
I opened the #519 bug.

**Reply by jkraskesis** [reply]
Thanks, the problem is solved!

**Reply by ilg-ul** [reply]
Great! Thank you for your bug report.

**Comment by wico-kulkarni** [comment]
Hi,

I too still see the same problem. The package update hangs on. Could someone help me please? What needs to be done to resolve this issue? 
Thanks,

FYI, I use Ubuntu 20.04 and Eclipse IDE for Embedded C/C++ Developers (Version: 2020-09 (4.17.0))

![Screenshot from 2023-05-23 10-54-58](https://github.com/eclipse-embed-cdt/eclipse-plugins/assets/132360840/58243920-7cf0-4f1f-b6a4-78600969e74c)

**Reply by ilg-ul** [reply]
Please update the plug-ins to the latest version.

**Comment by TommyMurphyTM1234** [comment]
I seem to be able to reproduce this too.
This is what I did.

1. Create a clean Zorin OS Lite 16.2 (Ubuntu 20.04 LTS derivative) VirtualBox VM with all the latest updates applied
2. Download and install Eclipse 2023.03 for Embedded C/C++ Developers using the installer here: https://www.eclipse.org/downloads/download.php?file=/oomph/epp/2023-03/R/eclipse-inst-jre-linux64.tar.gz
3. Launch Eclipse
4. _Window > Show View > Other..._,  search for _CMSIS Packs_, select it and clicl _Open_.
5. Click the Update the CMSIS packs definitions
6. I get this:

![2023-05-23 12 10 30](https://github.com/eclipse-embed-cdt/eclipse-plugins/assets/14550604/7629f1d0-c952-4f10-b039-492c37e72adf)

7. If I _Retry_ it just fails again after a while
8. I get the same results on Windows 10.
9. I can manually download the PDSC file from http://www.analog.com/media/en/engineering-tools/design-tools/AnalogDevices.ADuCM320_DFP.pdsc yielding this: 
[AnalogDevices.ADuCM320_DFP.zip](https://github.com/eclipse-embed-cdt/eclipse-plugins/files/11543201/AnalogDevices.ADuCM320_DFP.zip) but the browser does give a security warning (possibly because of the use of `http` rather than `https`?)
10. If I _Ignore_ (skip) that one then other CMSIS Packs, including other Analog Devices ones, download OK. Oh - sorry - spoke too soon: 

![2023-05-23 12 48 42](https://github.com/eclipse-embed-cdt/eclipse-plugins/assets/14550604/44136765-fdc7-4848-b1ec-8369fffbdc49)

And another one: 

![2023-05-23 12 56 59](https://github.com/eclipse-embed-cdt/eclipse-plugins/assets/14550604/b418219f-e599-4c45-bcb7-f86b8fb276f9)

BTW - a few other related observations (that perhaps should be logged separately?):

1. the CMSIS Packs view initially says "(no packages) - Press the Refresh button to register all available packages". However the relevant toolbar button in the view is not called "Refresh" but "Update the CMSIS packs definitions" so this seems a little confusing.
2. While the CMSIS Packs downloading is in progress the Cancel button doesn't seem to have any effect. (edit: OK - maybe it has not effect while it's stuck on downloading a specific pack but works "between" packs?)
3. Other nomenclature seems inconsistent - "CMSIS Pack(s)" versus "CMSIS pack(s)", "pack" versus "packages", "register" versus "refresh" (when it's actually download and register I think?) etc.
4. _Ignore_ and _Abort_ seem to have the same effect - skip the current pack which had a problem downloading?

**Comment by ilg-ul** [comment]
Hmmm... The mess with the CMSIS Packs seem to go worse and worse. :-(

> 1. the CMSIS Packs view initially says "(no packages) - Press the Refresh button to register all available packages". However the relevant toolbar button in the view is not called "Refresh" but "Update the CMSIS packs definitions" so this seems a little confusing.
> 3. Other nomenclature seems inconsistent - "CMSIS Pack(s)" versus "CMSIS pack(s)", "pack" versus "packages", "register" versus "refresh" (when it's actually download and register I think?) etc.

Agree, it is confusing. How would you rephrase those strings?

> 4. Ignore and Abort seem to have the same effect - skip the current pack which had a problem downloading?

Abort should stop the whole session, while Ignore should only skip the current package, and continue to process the remaining ones.

**Reply by TommyMurphyTM1234** [reply]
> Hmmm... The mess with the CMSIS Packs seem to go worse and worse. :-(
> 
> > 1. the CMSIS Packs view initially says "(no packages) - Press the Refresh button to register all available packages". However the relevant toolbar button in the view is not called "Refresh" but "Update the CMSIS packs definitions" so this seems a little confusing.
> > 2. Other nomenclature seems inconsistent - "CMSIS Pack(s)" versus "CMSIS pack(s)", "pack" versus "packages", "register" versus "refresh" (when it's actually download and register I think?) etc.
> 
> Agree, it is confusing. How would you rephrase those strings?

Let me create a separate ticket for the nomenclature issues.

> > 4. Ignore and Abort seem to have the same effect - skip the current pack which had a problem downloading?
> 
> Abort should stop the whole session, while Ignore should only skip the current package, and continue to process the remaining ones.

It doesn't.
_Abort_ has the same effect as _Ignore_ - it just skips the current problematic item and continues on to deal with the rest.
I'll also log this separately.

**Reply by unclebanjoman** [reply]
I'm having the same problem with CMSIS packs refresh it hangs forever in analogdevices.aducm335_dfp.pdsc.
I'm using Eclipse Oxygen.3 with latest gnu mcu installed from https://download.eclipse.org/embed-cdt/releases/4.7.2/p2/

**Reply by ilg-ul** [reply]
> CMSIS packs refresh it hangs forever ...

I just installed the 2024-12 release and did a full CMSIS Packs Refresh, so I cannot confirm that it hangs forever. 

```
java.net.SocketTimeoutException: Read timed out http://download.analog.com/tools/EZBoards/ADuCM355/Releases/AnalogDevices.ADuCM355_DFP.pdsc
Missing "/Users/ilg/Library/CMSIS-Packs/.cache/AnalogDevices.ADuCM355_DFP.1.0.4.pdsc", ignored by user request.
...
java.net.SocketTimeoutException: Read timed out https://download.analog.com/tools/EZBoards/ADuCM32x/Releases/AnalogDevices.ADuCM320_DFP.pdsc
Missing "/Users/ilg/Library/CMSIS-Packs/.cache/AnalogDevices.ADuCM320_DFP.1.1.1.pdsc", ignored by user request.
...
java.net.SocketTimeoutException: Read timed out https://download.analog.com/tools/EZBoards/ADuCM41x/Releases/AnalogDevices.ADuCM410_DFP.pdsc
Missing "/Users/ilg/Library/CMSIS-Packs/.cache/AnalogDevices.ADuCM410_DFP.1.7.0.pdsc", ignored by user request.
...
java.net.SocketTimeoutException: Read timed out http://download.analog.com/tools/EZBoards/ADSP-SC83x/Releases/AnalogDevices.ADSP-SC83x_FreeRTOS-OSAL.pdsc
Missing "/Users/ilg/Library/CMSIS-Packs/.cache/AnalogDevices.ADSP-SC83x_FreeRTOS-OSAL.1.0.0.pdsc", ignored by user request.
...
java.net.SocketTimeoutException: Read timed out http://download.analog.com/tools/EZBoards/ADSP-SC83x/Releases/AnalogDevices.ADSP-SC83x_DFP.pdsc
Missing "/Users/ilg/Library/CMSIS-Packs/.cache/AnalogDevices.ADSP-SC83x_DFP.2.0.0.pdsc", ignored by user request.
...

2025-02-03 10:09:23
Collecting devices...
Found 771 device(s), from 41 vendor(s).
Completed in 13ms.

2025-02-03 10:09:23
Collecting packs...
Found 8144 package version(s), from 70 vendor(s).
Completed in 81ms.
```

It is true that lots of packages failed after long timeouts and I had to manually ignore them, but this is a known issue.

It is the result of the broken Arm CMSIS Packs design, with a centralised index that Arm is not able to keep in sync with the vendor published packages.

Therefore the CMSIS Packs are deprecated and no longer recommended.

**Reply by unclebanjoman** [reply]
Thank you for your reply.
Currently I'm using Eclipse Oxygen.3 running on Debian 10 32-bit system with OpenJDK 1.8.0_332. Can I install the latest version on my system? And from which url repository?
For economic reasons I cannot use a more modern system.
Thanks again,
Max

**Reply by jonahgraham** [reply]
@unclebanjoman I don't know exactly which version of Eclipse dropped 32-bit support, but sadly you probably can't easily update.

**Reply by unclebanjoman** [reply]
Eclipse 32-bit support was dropped with Photon. Oxygen.3 was the last 32-bit version.
So I'm stuck I suppose..

**Comment by jessecheu** [comment]
I am having the same issue with the 2025-03 version 64bit Linux
Eclipse IDE for C/C++ Developers (includes Incubating components)
Version: 2025-03 (4.35.0)
Build id: 20250306-0812
I fully updated all the plugins.
